osquery/tables/smart/smart_drives.cpp

```diff
@@ -0,0 +1,219 @@
+    std::vector<hardwareDriver>& types) {
+  for (auto const& driver : storageDrivers) {
+    if (kSMARTExplicitDriverToDevice.find(driver) !=
+        kSMARTExplicitDriverToDevice.end()) {
+      types.push_back(kSMARTExplicitDriverToDevice.at(driver));
+    }
+  }
+}
+
+/// Utility function for traversing system devices.
+void walkDevices(std::function<bool(libsmartctl::Client& c,
+                                    const std::string& devname,
+                                    hardwareDriver* type)> handleDevF) {
+  if (getuid() || geteuid()) {
+    LOG(WARNING) << "Need root access for smart information";
+  }
+
+  QueryData results;
+  libsmartctl::Client c;
+
+  auto storageDrivers = getStorageCtlerClassDrivers();
+
+  std::vector<hardwareDriver> types;
+  getSmartCtlDeviceType(storageDrivers, types);
+
+  if (types.size() > 1) {
+    LOG(WARNING) << "Found more than 1 hardware storage controller:" << types
+                 << "; only handling the first";
+    return;
+  }
+
+  hardwareDriver* type{nullptr};
+  if (types.size() == 1) {
+    type = &(types[0]);
+  }
+
+  auto blkDevices = SQL::selectAllFrom("block_devices");
+  for (const auto& device : blkDevices) {
+    if (device.find("size") != device.end() && device.at("size") != "0" &&
+        device.at("size") != "" && device.find("name") != device.end()) {
+      if (handleDevF(c, device.at("name"), type)) {
+        break;
+      };
+    }
+  }
+}
+
+QueryData genSmartInfo(QueryContext& context) {
+  QueryData results;
+  /* hwInfo is for tracking info retrieve with an explicit HW controller.  It is
+   * indexed by serial_number, since that's how you correlate the data with
+   * auto-detect retrieved SMART info. */
+  std::map<std::string, Row> hwInfo;
+
+  bool found{false};
+  walkDevices([&](libsmartctl::Client& c,
+                  const std::string& devname,
+                  hardwareDriver* type) {
+    // Get auto info..
+    auto resp = c.getDevInfo(devname, "");
+    if (resp.err != NOERR) {
+      LOG(INFO) << "There was an error retrieving drive information: "
+                << libsmartctl::errStr(resp.err);
+
+    } else {
+      resp.content["device_name"] = devname;
+      results.push_back(std::move(resp.content));
+    }
+
+    // Get info via HW controllers if not found yet
+    if (found) {
+      return false;
+    }
+
+    if (type == nullptr) {
+      found = true;
+      return false;
+    }
+
+    if (devname.substr(devname.length() - 1).find_last_of("0123456789") !=
+        std::string::npos) {
+      return false;
+    }
+
+    for (size_t i = 0; i <= type->maxID; i++) {
+      std::string fullType = type->driver + std::to_string(i);
+
+      auto cantId = c.cantIdDev(devname, fullType);
+      if (cantId.err != NOERR) {
+        LOG(INFO) << "Error while trying to identify device: "
+                  << libsmartctl::errStr(cantId.err);
+        continue;
+      }
+      // If device is not identifiable, the type is invalid, skip
```

Comment:
nit: there's some inconsistent spacing between clauses of code, can those be cleaned up?

Reply:
yep.